### PR TITLE
listener: fix inplace update with conn balancer

### LINKED ISSUE #12686: Listener Crash on Start after enabling exact_balance
Listener Crash on Start after enabling exact_balance

_NOTE: emailed envoy-security@ and was asked to create an issue by @mattklein123 - I'm in the process of extracting more information, wanted to kick-off the discussion in the interim._

Envoy crashes on startup with exact_balance configuration enabled.

       "connection_balance_config": {

        "exact_balance": {}

       }

Envoy version:

       http_archive(
        name = "envoy",
        sha256 = "38bd41e5229532abbeccba7b87d80c8664e915ec6f780a60a6b7ef1818458b5a",
        strip_prefix = "envoy-1.15.0",
        urls = ["https://github.com/envoyproxy/envoy/archive/v1.15.0.tar.gz"],
      )

backtrace:
 
```
(gdb) bt    
#0  Envoy::Network::ExactConnectionBalancerImpl::pickTargetHandler (this=0x556dcafbe900) at external/envoy/source/common/network/connection_balancer_impl.cc:31
#1  0x0000556dc59eb6a3 in Envoy::Server::ConnectionHandlerImpl::ActiveTcpListener::onAcceptWorker (this=0x556dca71d300, socket=..., hand_off_restored_destination_connections=<optimized out>, rebalanced=<optimized out>) at external/envoy/source/server/connection_handler_impl.cc:347
#2  0x0000556dc5a0618d in Envoy::Network::ListenerImpl::listenCallback (fd=<optimized out>, remote_addr=<optimized out>, remote_addr_len=<optimized out>, arg=<optimized out>) at external/envoy/source/common/network/listener_impl.cc:78
#3  0x0000556dc5e6216b in listener_read_cb (fd=55, what=<optimized out>, p=0x556dcc113ad0) at /root/.cache/bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/32/execroot/gateway/external/com_github_libevent_libevent/listener.c:437
#4  0x0000556dc5e5ee80 in event_persist_closure (ev=<optimized out>, base=0x556dc8d6f340) at /root/.cache/bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/32/execroot/gateway/external/com_github_libevent_libevent/event.c:1639
#5  event_process_active_single_queue (base=base@entry=0x556dc8d6f340, max_to_process=max_to_process@entry=2147483647, endtime=endtime@entry=0x0, activeq=<optimized out>) at /root/.cache/bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/32/execroot/gateway/external/com_github_libevent_libevent/event.c:1698
#6  0x0000556dc5e5f4ff in event_process_active (base=0x556dc8d6f340) at /root/.cache/bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/32/execroot/gateway/external/com_github_libevent_libevent/event.c:1799
#7  event_base_loop (base=0x556dc8d6f340, flags=0) at /root/.cache/bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/32/execroot/gateway/external/com_github_libevent_libevent/event.c:2041
#8  0x0000556dc59e70cd in Envoy::Server::WorkerImpl::threadRoutine (this=0x556dc8db5310, guard_dog=...) at external/envoy/source/server/worker_impl.cc:133
#9  0x0000556dc5f20475 in std::function<void ()>::operator()() const (this=<optimized out>) at /usr/include/c++/7/bits/std_function.h:706
#10 Envoy::Thread::ThreadImplPosix::ThreadImplPosix(std::function<void ()>, absl::optional<Envoy::Thread::Options> const&)::{lambda(void*)#1}::operator()(void*) const (arg=<optimized out>, __closure=0x0) at external/envoy/source/common/common/posix/thread_impl.cc:49
#11 Envoy::Thread::ThreadImplPosix::ThreadImplPosix(std::function<void ()>, absl::optional<Envoy::Thread::Options> const&)::{lambda(void*)#1}::_FUN(void*) () at external/envoy/source/common/common/posix/thread_impl.cc:51
#12 0x00007f33559246db in start_thread (arg=0x7f333650f700) at pthread_create.c:463
#13 0x00007f335564da3f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

code link: https://github.com/envoyproxy/envoy/blob/master/source/common/network/connection_balancer_impl.cc#L31

min_connection_handler is a nullptr as there aren't any BalanceConnectionHandler handlers registered when this balancing occurs.

```
// printout of handlers_ (empty list)
  handlers_ = {
    <std::_Vector_base<Envoy::Network::BalancedConnectionHandler*, std::allocator<Envoy::Network::BalancedConnectionHandler*> >> = {
      _M_impl = {
        <std::allocator<Envoy::Network::BalancedConnectionHandler*>> = {
          <__gnu_cxx::new_allocator<Envoy::Network::BalancedConnectionHandler*>> = {<No data fields>}, <No data fields>}, 
        members of std::_Vector_base<Envoy::Network::BalancedConnectionHandler*, std::allocator<Envoy::Network::BalancedConnectionHandler*> >::_Vector_impl: 
        _M_start = 0x0, 
        _M_finish = 0x0, 
        _M_end_of_storage = 0x0
      }
    }, <No data fields>} 
```

In-order to see if we could mitigate the crash, we were returning the handler passed as the sole argument to ExactConnectionBalancerImpl::pickTargetHandler. It resolves the crash on startup, but, then we don't achieve exact balancing consistently. Below are stats dumped for the worker's on two of our staging clusters deployed in different zones.

example:
# a case where it was working after the patch
listener.0.0.0.0_8080.worker_0.downstream_cx_active: 810
listener.0.0.0.0_8080.worker_0.downstream_cx_total: 1092
listener.0.0.0.0_8080.worker_1.downstream_cx_active: 811
listener.0.0.0.0_8080.worker_1.downstream_cx_total: 1072
listener.0.0.0.0_8080.worker_2.downstream_cx_active: 810
listener.0.0.0.0_8080.worker_2.downstream_cx_total: 1112
listener.0.0.0.0_8080.worker_3.downstream_cx_active: 811
listener.0.0.0.0_8080.worker_3.downstream_cx_total: 1086
listener.0.0.0.0_8080.worker_4.downstream_cx_active: 811
listener.0.0.0.0_8080.worker_4.downstream_cx_total: 1096

# a case where it was not working (different cluster)
listener.0.0.0.0_8080.worker_0.downstream_cx_active: 1030
listener.0.0.0.0_8080.worker_0.downstream_cx_total: 2236
listener.0.0.0.0_8080.worker_1.downstream_cx_active: 452
listener.0.0.0.0_8080.worker_1.downstream_cx_total: 1381
listener.0.0.0.0_8080.worker_2.downstream_cx_active: 125
listener.0.0.0.0_8080.worker_2.downstream_cx_total: 875
listener.0.0.0.0_8080.worker_3.downstream_cx_active: 252
listener.0.0.0.0_8080.worker_3.downstream_cx_total: 1054
listener.0.0.0.0_8080.worker_4.downstream_cx_active: 1812
listener.0.0.0.0_8080.worker_4.downstream_cx_total: 3744

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:
By sharing the connection balancer among listener configs.
Signed-off-by: Yuchen Dai <silentdai@gmail.com>

Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Runtime guard:]
Fixes #12686 
[Optional Deprecated:]

### CODE REVIEW

File: test/integration/xds_integration_test.cc [hunk 1]
```diff
@@ -266,6 +266,10 @@ class LdsInplaceUpdateHttpIntegrationTest
     config_helper_.addListenerFilter(tls_inspector_config);
     config_helper_.addSslConfig();
     config_helper_.addConfigModifier([](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+      bootstrap.mutable_static_resources()
+          ->mutable_listeners(0)
+          ->mutable_connection_balance_config()
+          ->mutable_exact_balance();
```
Comment (mattklein123) [inline test/integration/xds_integration_test.cc:272]: Can you make this a setup param and test this both ways (default off is fine just want to have at least 1 test where we have the other config.)

**Comment by wozz** [comment]
was this fixed by another PR? or can we reopen this?

**Comment by mattklein123** [comment]
@lambdai can you finish this fix? Otherwise I can do it this week.

**Comment by lambdai** [comment]
Sorry… updating this pr.

File: docs/root/version_history/current.rst [hunk 2]
```diff
@@ -28,8 +28,9 @@ Minor Behavior Changes
 * http: fixed the 100-continue response path to properly handle upstream failure by sending 5xx responses. This behavior can be temporarily reverted by setting `envoy.reloadable_features.allow_500_after_100` to false.
 * http: the per-stream FilterState maintained by the HTTP connection manager will now provide read/write access to the downstream connection FilterState. As such, code that relies on interacting with this might
   see a change in behavior.
-* logging: add fine-grain logging for file level log control with logger management at administration interface. It can be enabled by option `--enable-fine-grain-logging`.
-* logging: change default log format to `"[%Y-%m-%d %T.%e][%t][%l][%n] [%g:%#] %v"` and default value of :option:`--log-format-prefix-with-location` to `0`.
+* listener: fixed crash at listener inplace update when connetion load balancer is set.
```
Comment (mattklein123) [inline docs/root/version_history/current.rst:31]: Can you move this to the bug fix section?

File: test/integration/xds_integration_test.cc [hunk 3]
```diff
@@ -356,11 +362,18 @@ class LdsInplaceUpdateHttpIntegrationTest
...
+    codec_client_after_config_update->close();
+  }
+
   std::unique_ptr<Ssl::ContextManager> context_manager_;
   Network::TransportSocketFactoryPtr context_;
   testing::NiceMock<Secret::MockSecretManager> secret_manager_;
   Network::Address::InstanceConstSharedPtr address_;
-};
+  bool use_default_balancer_{false};
+}; // namespace
```
Comment (mattklein123) [inline test/integration/xds_integration_test.cc:376]: ?

File: test/server/connection_handler_test.cc [hunk 4]
```diff
@@ -60,14 +60,18 @@ class ConnectionHandlerTest : public testing::Test, protected Logger::Loggable<L
...
+        uint32_t tcp_backlog_size = ENVOY_TCP_BACKLOG_SIZE,
+        Network::ConnectionBalancerSharedPtr connection_balancer = nullptr)
         : parent_(parent), socket_(std::make_shared<NiceMock<Network::MockListenSocket>>()),
           socket_factory_(std::move(socket_factory)), tag_(tag), bind_to_port_(bind_to_port),
           tcp_backlog_size_(tcp_backlog_size),
           hand_off_restored_destination_connections_(hand_off_restored_destination_connections),
           name_(name), listener_filters_timeout_(listener_filters_timeout),
           continue_on_listener_filters_timeout_(continue_on_listener_filters_timeout),
-          connection_balancer_(std::make_unique<Network::NopConnectionBalancerImpl>()),
+
```
Comment (mattklein123) [inline test/server/connection_handler_test.cc:71]: nit: delete newline

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small comments.

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks!

File: source/server/listener_impl.cc [hunk 5]
```diff
@@ -325,6 +325,7 @@ ListenerImpl::ListenerImpl(ListenerImpl& origin,
       listener_filters_timeout_(
           PROTOBUF_GET_MS_OR_DEFAULT(config, listener_filters_timeout, 15000)),
       continue_on_listener_filters_timeout_(config.continue_on_listener_filters_timeout()),
+      connection_balancer_(origin.connection_balancer_),
```
Comment (lambdai) [inline source/server/listener_impl.cc:328]: mark

**Pull request merged**